Update ECC public key check to use openssl3+ APIs

ColtonWilley opened the pull request:
# Goal
Refactor s2n_ecc_check_key() to use EVP APIs instead of direct EC_KEY APIs for public key validation.

## Why
Existing methods to check EC public key are not provider enabled, breaking attempts to use an alternate crypto provider. 

## How
-Added new EVP-based s2n_ecc_check_key() implementation gated by EVP_APIS_SUPPORTED
-Uses EVP_PKEY_public_check() for key validation
-Returns FIPS-specific error code when s2n_is_in_fips_mode() is true
-Preserved EC_KEY fallback for older libcryptos

## Testing
All existing ECC tests pass.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ColtonWilley commented on crypto/s2n_ecc_evp.c:
```diff
@@ -206,21 +209,44 @@ static int s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_
 
-#ifdef S2N_LIBCRYPTO_SUPPORTS_EC_KEY_CHECK_FIPS
+    if (s2n_is_in_fips_mode()) {
+        RESULT_ENSURE(result == 1, S2N_ERR_ECDHE_INVALID_PUBLIC_KEY_FIPS);
```
I apologize for the delay, I got busy with some other work. I only did it this way to preserve return behavior. The new model just does a key check, if FIPS is enabled on the underlying library then it becomes a FIPS check. On the offchance existing applications depended on this behavior I tried to preserve it, but I am happy to do it either way. If you would like to see a particular change here let me know and I can make it.

xnox replied:
piblic check key is supported in both fips and non-fips modes.

basic checks are performed by both modes, and it is prudent for any tls implementations to call this always.

in fips mode can some additional checks might be performed, which often amount to computing public key twice for no additional safety gains.

I would recommend to call this check always regardless of fips or non-fips mode.

jmayclin commented:
Hi @ColtonWilley !

Thanks for the PR, but I wanted to give you a heads up that our review won't be prompt because we're currently figuring out long-term OpenSSL support https://github.com/aws/s2n-tls/issues/5783, and we want to minimize OpenSSL related changes until we have an answer on that.

The recommended libcrypto to use with s2n-tls is AWS-LC. OpenSSL support is something we do to try and meet our customers where they're at, but as OpenSSL and AWS-LC (originally a fork of BoringSSL) continue to diverge, we're finding that the maintenance cost continues to increase, especially with OpenSSL 4.0 support requests on the horizon.

So while I have you on the line, I'm wondering if you have any thoughts on the impact of s2n-tls interning (statically linking and symbol prefixing) AWS-LC by default, as this is one of the potential solutions we're exploring.

ColtonWilley commented:
Apologies for the delays,

@jmayclin I think for your own needs that might make sense. Most of the use cases we are seeing is from customers wanting FIPS certified crypto usage in their applications. In that sense such a decision would make it harder for users of s2n to get on a path to certified crypto. In addition, as PQC comes down the line if you intern the crypto you will now be responsible for being up to date, or s2n users wont have up to date paths for PQC. 

I think it does make s2n less crypto agile, but I also fully understand the burden of supporting both as a maintainer. I think it should be noted we are seeing more and more customers wanting s2n + FIPS, so as we see breakage you would be getting patches like this along the way. 

For this PR I think testing is ok going either way, if you have a preference please let me know.

ColtonWilley committed:
Update ECC public key check to use openssl3+ APIs

The low-level EC_KEY APIs (EVP_PKEY_get1_EC_KEY, EC_KEY_check_key) are
deprecated in OpenSSL 3.0+. Replace them with EVP_PKEY_public_check, which
performs the required NIST-curve public key validation directly on the
EVP_PKEY.

The pre-3.0 path is unchanged in behavior: it still extracts the EC_KEY and
uses EC_KEY_check_fips in FIPS mode when the libcrypto supports it (e.g.
AWS-LC), falling back to EC_KEY_check_key otherwise.

Per review feedback, the OpenSSL 3.0+ path does not raise the FIPS-specific
error: EVP_PKEY_public_check is the same call regardless of FIPS mode (the
FIPS provider, when active, performs the appropriate checks), so a distinct
FIPS error code is not meaningful there. The EC_KEY_free cleanup func is now
gated to the pre-3.0 paths that still need it.
